strong_params

why-el opened the pull request:
https://community.openproject.org/work_packages/21724/activity

Pending these PR merges:
- [x] https://github.com/finnlabs/openproject-backlogs/pull/179
- [x] https://github.com/finnlabs/openproject-announcements/pull/18/
- [x] https://github.com/finnlabs/openproject-costs/pull/173
- [x] https://github.com/finnlabs/reporting_engine/pull/56
- [x] https://github.com/finnlabs/openproject-global_roles/pull/44
- [x] https://github.com/finnlabs/openproject-impermanent_memberships/pull/50
- [x] https://github.com/finnlabs/openproject-landing_page/pull/41
- [x] https://github.com/finnlabs/openproject-dtag_customizing/pull/38

Commit:
Refactor to allow for strong params

We no longer need attr_accessible.

Commit:
Add responsible disclosure section to the README

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://why-el-sec-fixes.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://why-el-sec-fixes.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

Commit:
Delete Redmine's safe_attributes implementation

It was an early implementation of strong_parameters, more or less.
Going forward, we should strive to make the whitelisting of attributes
a controller responsibility only.

oliverguenther commented on app/controllers/account_controller.rb:
```diff
@@ -85,7 +85,7 @@ def lost_password
     else
       if request.post?
-        user = User.find_by_mail(params[:mail])
+        user = User.find_by(mail: params[:mail])
```
Note that there is a explicit method `find_by_mail` for users, this is not one of those deprecated dynamic finders.

why-el replied:
Good catch sir. Looks like that method was finding them case-insensitive, so I will restore it.

Commit:
Restore own User#find_by_mail impl, as it was doing more than just find

Commit:
Fix legacy and cucumber specs failing because of strong params

why-el commented:
@oliverguenther or @basiszwo a quick review here? Then we can merge.

basiszwo commented:
looks good to me. great work! did you run this against the community edition testsuite?

basiszwo commented:
I guess this is wip, right?

why-el commented:
Yes, sorry. Lots of plugins went berserk.

Commit:
Ensure versions params play well with plugins

basiszwo commented:
went through the plugins they look good to me. 
can you please rebase this branch?

why-el commented:
Will do.

why-el commented:
@basiszwo this is rebased. Can we move on with it? Since the failing cukes are not caused by this?

basiszwo commented:
ok, I take over and provide an environment to check stuff

why-el commented:
@basiszwo I added multiple plugins to this, basically all ones that we making `myproject` fail, can we check it?

why-el commented:
@ulferts wanna take a look here?

ulferts commented on app/models/mail_handler.rb:
```diff
@@ -161,8 +161,8 @@ def receive_issue
     issue = WorkPackage.new(author: user, project: project)
-    issue.safe_attributes = issue_attributes_from_keywords(issue)
-    issue.safe_attributes = { 'custom_field_values' => custom_field_values_from_keywords(issue) }
+    issue.attributes = issue_attributes_from_keywords(issue)
```
@why-el please note that `safe_attributes=` has some more behaviour on [Issues/WorkPackages](https://github.com/opf/openproject/blob/dev/app/models/work_package.rb#L518). I haven't yet looked into whether that will present a problem, but it should be double checked. It e.g. checks whether a new status is allowed. Taking the status as an example, the check within safe attributes is IMO misplaced, a validation would be better (as long as one wants to stick to the model layer), but the check exists now and we have to make sure, that a replacement exists.

Additionally, I am undecided whether we need to wrap the call to `issue_attributes_from_keywords` into some permit logic. We are accepting input from an e-mail here. While we are parsing the mail and thus should be able to control the parameters, we would have to maintain the logic about what is allowed for whom within the `MailHandler` which is not a good place as the logic is also required elsewhere.

why-el replied:
Yep, I paused there for a minute but specs reported no failures, so I assumed the change is ok. Maybe we can add a test case for statues?

For `issue_attributes_from_keywords`, it looks fine to me. We are constructing the hash ourselves (so essentially the same as what strong_params provides).

ulferts replied:
After taking a deeper look into the current `WorkPackage#safe_attributes=` implementation I think that the change does alter the behaviour. The method implements some business logic:
1) Some attributes are removed when the work package has a parent
2) The new status is checked
3) The parent_id is removed on some cases.

In my opinion the above problem and the one with with `issue_attributes_from_keywords` boils down to an architectural decision. Is permitted_params only responsible for statically whitelisting some attributes for mass assignment or is there business logic involved? Currently it is the later as decisions are made depending e.g. on the user and for some parameters also on the state of the object the attributes are to be assigned to. 

So a simple solution might be to 
1) move the business logic in `safe_attributes=` to `PermittedParams`. You have already done that for other models.
2) Wrap the response from `issue_attributes_from_keywords` in the same permit logic to benefit from the business logic. 
This solution might be good enough for now. 

However, I find this to be a suboptimal solution. Doing that we are spreading the business logic (e.g. Model, PermittedParams, Controller, Mailer). 
We need some object, that contains the whole of the business logic. There already exist some service objects so they might be the container I am looking for. DCI or some other POR business layer between controller and model might also do the trick.

why-el replied:
Yes, `PermittedParams` as defined by OP gets access to the user instance, so we can do permission logic there, which is an improvement over doing it in the model (most of which I deleted). 

The `safe_attributes=` is our own implementation, not the one from Redmine, so I left it there as it is doing more work (but I deleted where it was used, because i conflicted it with the Redmine one), so how about I restore it there? This is even simpler than your solutions (which are good so we can do them in a different package).

ulferts replied:
The `safe_attributes=` on WorkPackages is overwriting the one from `Redmine::SafeAttributes`. Restoring the calls to it will not work because the method is not working after your removal of the module. The method calls `delete_unsafe_attributes` which is defined in `Redmine::SafeAttributes`. So the `safe_attributes=` method on `WorkPackage` has to be removed and we need to find an alternate place for the logic within it.  

There is another method within `WorkPackage` called `update_by!`/`update_by` which could be a place for the business logic. One notable thing is though, that said method was circumventing the business logic defined in `safe_attributes=` the whole time. Not quite sure where that leaves us. As far as I can tell, the `update_by` method is used in the more important places and there haven't been any bug reports regarding behaviour on that paths. So we might be able to delete the method altogether without moving the logic.

Another deeper look at what is in `safe_attributes=` will tell you that. 

If we decide not to delete (parts) of the method, another place to put the logic would be the `UpdateWorkPackageService` by the way, which already exists.

ulferts commented on app/models/mail_handler.rb:
```diff
@@ -188,8 +188,8 @@ def receive_issue_reply(issue_id)
-    issue.safe_attributes = issue_attributes_from_keywords(issue)
-    issue.safe_attributes = { 'custom_field_values' => custom_field_values_from_keywords(issue) }
+    issue.attributes = issue_attributes_from_keywords(issue)
+    issue.attributes = { 'custom_field_values' => custom_field_values_from_keywords(issue) }
```
same as above

ulferts commented on app/models/work_package.rb:
```diff
@@ -508,7 +502,6 @@ def estimated_hours=(h)
-  # <<< issues.rb <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
 
   # Safely sets attributes
   # Should be called from controllers instead of #attributes=
```
The method below (`safe_attributes=`) should also be removed

why-el replied:
Which one again? (you are referring to a code comment)

ulferts replied:
The `safe-attributes=` [method](https://github.com/opf/openproject/blob/dev/app/models/work_package.rb#L518)

ulferts commented on app/workers/copy_project_job.rb:
```diff
@@ -77,7 +77,7 @@ def create_project_copy(source_project,
       parent_id = target_project_params[:parent_id]
       target_project = Project.new.tap do |p|
-        p.safe_attributes = target_project_params
+        p.attributes = target_project_params
```
@why-el this allows to set more attributes now, than it used to. Please double check, whether this is a problem.

why-el replied:
It's not, since we are passing `permitted_params.project` to it.

ulferts commented on lib/copy_model.rb:
```diff
@@ -37,7 +37,7 @@ module InstanceMethods
       with_model(from_model) do |model|
         # clear unique attributes
-        self.safe_attributes = model.attributes.dup.except(*(Array(self.class.not_to_copy).map(&:to_s)))
+        self.attributes = model.attributes.dup.except(*(Array(self.class.not_to_copy).map(&:to_s)))
```
@why-el this again has way less security than it used to.

why-el replied:
How so?

jonasheinrich commented on app/models/permitted_params.rb:
```diff
@@ -262,6 +236,155 @@ def wiki_content
+      whitelist[:custom_field_values] = params[:project][:custom_field_values]
+    end
+
+    whitelist
```
isn't this call superfluous since `tap` already returns the object?

why-el replied:
Nope, the `break` will force a `nil` return.

jonasheinrich replied:
Ok, didn't see that. What about `break whitelist`? Not sure if this is better but looks more compact.
Otherwise you don't need the `tap` at all, do you? Wouldn't an `unless` have the same meaning?

oliverguenther replied:
Just chiming in due to mail notification, I prefer Waels more explicit approach of returning whitelist after the tap. I find it more concise than forcing break to return it, too. There's no wondering what's happening when you explicitly return it ;)

why-el replied:
Using the tapped into object inside its own tap is confusing, at least to my lazy eyes. :D 

And sure, if you have a more compact way I'd take a look :) I prefer explicit returns, this way if more processing of params is needed, people will not need to read my tap closely to do their job, less errors, etc. This is a trivial piece of code though so not sure if it is worth your time.

oliverguenther replied:
If you return it explicity, you might as well remove the `tap` code since that's essentially unnecessary.

why-el replied:
Yep agreed, I think the tap survived earlier code, but I will delete it now.

ulferts commented on app/models/permitted_params.rb:
```diff
@@ -262,6 +236,155 @@ def wiki_content
+  def membership
+    params.require(:membership).permit(*self.class.permitted_attributes[:membership])
+  end
+
```
We need unit tests for those added methods.

why-el replied:
Is that worth it currently? There are all pretty basic except for a one or two. 

Stefan and I plan to move these to their respective controllers, and after that controller or feature specs will suffice, except when there is small authorization logic then we can test that as well.

ulferts replied:
I think they are very fast to write and execute and prevent unwanted change.

I personally dislike the idea of moving those to the controllers as it will just bloat the controllers and will make reuse quite hard (e.g. between a standard controller and an API one). I'd very much like to sit down with @basiszwo, you and whoever likes to (or is not fast enough too run away in time) and come up with a basic structure for those concerns. I think we should have some guidelines in place so that we don't have to discuss this on every other PR.

why-el replied:
Sure thing, I can write them. And sounds good for the discussion.

Commit:
Fix issue with versions/new being not accessible

Commit:
Add specs for new strong_params methods

why-el commented:
@oliverguenther can you take a quick pass at the changes in plugins? After that we can merge.

oliverguenther commented:
@why-el I've only had a quick look at the code (did not test the configuration with changed plugins) since my priorities lie elsewhere. That being said, the changes appear clear and look fine to me. If you've tested the other configuration and @ulferts agrees on the changes on this PR itself, I'd say go for it.

why-el commented:
Ok great. Will need a rebase so I will do that later today.

lindenthal commented:
@why-el: what is the status here?

why-el commented:
Merging today.

Commit:
Merge branch 'dev' into sec-fixes
